ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -252,6 +252,63 @@ private class ContractsFetch(
     }
   }
 
+  def fetchAndRefreshCache(
+      jwt: Jwt,
+      ledgerId: LedgerApiDomain.LedgerId,
+      ledgerEnd: Terminates.AtAbsolute,
+      optOffsetToUpdate: Option[domain.Offset],
+  )(implicit
+      ec: ExecutionContext,
+      mat: Materializer,
+      lc: LoggingContextOf[InstanceUUID with RequestID],
+      metrics: HttpJsonApiMetrics,
+  ): ConnectionIO[List[BeginBookmark[domain.Offset]]] = {
+    import sjd.q.queries
+    import cats.syntax.traverse._
+
+    val ledgerOffset = ledgerEnd.toDomain
+    debugLogActionWithMetrics(
+      s"cache refresh for templates older than ledger offset: $ledgerOffset",
+      metrics.Db.warmCache,
+    ) {
+      for {
+        allOffsets <- queries.allOffsetsInformation(
+          ledgerOffset.unwrap,
+          optOffsetToUpdate.map(_.unwrap),
+        )
+        filteredTemplateInfoAndOffset =
+          allOffsets.map { case (_, partyOffsetNonEmpty) =>
+            partyOffsetNonEmpty.map { case (partyId, offset, packageId, moduleName, entityName) =>
+              type L[a] = (a, domain.Offset)
+              (
+                ContractTypeId.Template(packageId, moduleName, entityName),
+                domain.Party.subst[L, String](domain.Offset.tag.subst((partyId, offset))),
+              )
+            }
+          }.toList
+        partiesOffsets = filteredTemplateInfoAndOffset.map(_.toSet.map(_._2))
+        templateIds = filteredTemplateInfoAndOffset.flatMap(_.toSet.map(_._1))
+
```

Comment:
If you have the signature suggested above you should be able to do something like

```suggestion
        val templateOffsets: Map[ContractTypeId.Template, Map[domain.Party, domain.Offset]]
 = allOffsets.map { 
          case ((pac, mod, ent), partyOffsets) =>
            val k = ContractTypeId.Template(pac, mod, ent)
            val v = partyOffsets.map { case (pty, off) => (domain.Party(pty), domain.Offset(off))
            (k, v)
        }
```

Reply:
to be honest I think the types that are now are the best suitable for the cases we require later on.... I can discuss with you them.... changing to that type will require still the call to the toSet for the parties and toMap after the transformation of type (String, String) -> (Party, Offset).... so no benefit again doing this